### DISCUSSION TITLE
What is the difference between Rollback & MigrateDown in Fluent Migrator?

### DISCUSSION THREAD

**Discussion by virgilp** [fluentmigrator/fluentmigrator#1518] [Q&A] [answered]
**Describe the question**
The documentation is really unclear what's the difference, and other questions I found don't address that (beyond superficial differences like "migrate:down" without version migrates down all migrations whereas "rollback" without version rolls back one migration).

But, what is the difference between the concept of "rollback" and "migrate down"? If you specify a version, is there any difference? From code, I can tell that:

- If you migrate down all the way, "rollback" seems to delete the version table whereas "migrate:down" does not. Seems like a detail, and possibly a bug/ something that was overlooked.
- The check of which migrations to apply (down) is implemented rather differently, but to the best of my understanding, the net effect is the same. 
- After running the "down" migrations, "migrate:down" will also invoke ApplyProfiles() (which is: `foreach (var profile in Profiles) runner.Up(profile);` )

I'm not even sure if the whole ApplyProfiles() thing is intentional or a bug (given that it does `runner.Up()` in a `migrate:down` context). But, this seems to be the intended difference, if any? 

**Documentation Pages You've Read So Far**
https://fluentmigrator.github.io/articles/runners/runner-console.html#--task--tvalue-optional and stackoverflow questions that I found.

**Expected benefit**
If I got the answer to my question, I could decide whether to use `migrate:down --version` or `rollback:toversion --version`

**Information (please complete the following information):**
 - OS: [ Multiple ]
 - Platform: [ .NET 5 ]
 - FluentMigrator version [ 3.2.17 ]
 - FluentMigrator runner [ All - I do use "in-process runner" but it doesn't really matter]
 - Database Management System [ SQLServer ]
 - Database Management System Version [ SQL Server 2019 Enterprise Edition ]

**Additional context**
N/A

**Comment by jzabroski** [comment, marked as answer]
`rollback` rolls back to a state where its as if nothing happened.  Rolling back all the way deletes the version table, because if you rollback everything, you should... rollback everything.  This can be desirable in integration test scenarios where you are trying to run data access layer tests against your database, but do so in a manner where you rollback all changes.  One reason you might do this is if your test database is an ambient resource, and you continuously switch branches or update the HEAD of your branch.

`migrate:down` simply does an incremental rollback of specific migrations.  It could be that you discovered a particular migration was a bad idea, but you don't want to rollback the entire release.  Alternatively, you might want to debug the impact of applying a specific migration out-of-order, and so you roll it back, and then re-apply it, but to the end of the migration sequence.  Since migrate:down isn't a pure operation, there is some risk that you're at the mercy of the developer who wrote the Down migration.  If they left the Down migration blank, then it's a no-op.

Longer term, I would like to simplify this so that --version is replaced with a range parameter.  See #1033
